packages/editor/src/store/selectors.js

```diff
@@ -839,7 +839,7 @@ export function getEditedPostPreviewLink( state ) {
 	}
 
 	let previewLink = getAutosaveAttribute( state, 'preview_link' );
-	if ( ! previewLink ) {
+	if ( ! previewLink || 'draft' === getCurrentPost( state ).status ) {
```

Comment:
I'd also consider adding a reference to the Issue for context.

Reply:
hi @getdave ! Thanks for the feedback! 

One thing I noticed about Draft posts is that they do not update the Autosave record in the Database.

![Screen Shot 2021-12-10 at 2 40 45 PM](https://user-images.githubusercontent.com/3419200/145633739-d35e3d0c-d537-4760-9caf-f38d40535f1c.png)

In the screenshot we can see the records for the post. The first row is the Autosave that was generated when the Post was published and previewed. The other 3 records are changes to the post that were made after switching to Draft mode. The previewLink seems to be generated based on the autosave record, so I thought adding this status check could be a good way to generate a valid previewLink. 

Another observation is that after switching to Draft, and reloading the page. The autosave record gets deleted from the Database, and everything starts working fine again. 

So based on these findings, I think other alternatives are:
a) Ensure that the autosave record gets updated also for Drafts, not only for Published.
b) Delete the autosave record asynchronously when switching to draft.
c) The getAutosaveAttribute is marked as deprecated since 5.6, so upgrading to the recommended getAutosave() would be worth exploring too.

As I am a new contributor, I don't have much understanding about the implications of any of these alternatives, or if we are talking about updates to the Wordpress core, so any advice is appreciated.

Thanks!

Reply:
> hi @getdave ! Thanks for the feedback!

You're welcome!

> The previewLink seems to be generated based on the autosave record

Can you validate this? What code path is used and where? I'd like to understand more about what actually generates the preview link?

> ...so I thought adding this status check could be a good way to generate a valid previewLink.

It does _seem_ to but like you I'm also not 100% clear on this part of the system. When does the auto save record get created? Some possible avenues to explore

- does an autosave get generated when working on a post that has not been published (i.e. has always been a draft)?
- does an autosave get generated when working on a post that has been reverted from published to draft?
- what code paths force the creation of a new autosave record? 
- is the autosave record "stale" when switching to draft? 

Depending on the answers to the above, I wonder whether we are missing generation of an autosave record when switching from Publish to Draft status? 

> Ensure that the autosave record gets updated also for Drafts, not only for Published.

This is a good question. I'd need to understand more about the purpose of autosaves and their expected behaviour. Initially it would seem logical that autosaves would be generated for draft posts. After all if you've changed the post to "draft" status then you still want you changes to be autosaved right?

> Delete the autosave record asynchronously when switching to draft.

I think the above above with updating the record sounds better.

> The getAutosaveAttribute is marked as deprecated since 5.6, so upgrading to the recommended getAutosave() would be worth exploring too.

Worth a look. It would be nice to upgrade the code path to use non-deprecated methods. I would probably do this in a separate PR.

> As I am a new contributor, I don't have much understanding about the implications of any of these alternatives, or if we are talking about updates to the Wordpress core, so any advice is appreciated.

I also don't have much context on autosaves so we'll have to dig into this together. I'll see if I can get some more advice on this as well.

Thanks for all your work here 🙇‍♂️

Reply:
hi @getdave,

Answering your questions:

_does an autosave get generated when working on a post that has not been published (i.e. has always been a draft)?_
No, autosave records get generated only for published posts.

_does an autosave get generated when working on a post that has been reverted from published to draft?_
No, an autosave record is not generated in this case. The autosave record that was created when the post was published remains in the Database. Subsequent updates to the draft post do not update the autosave record.

_what code paths force the creation of a new autosave record?_
New autosave record gets created only for published post, when clicking on the 'Preview in new tab' option from the dropdown. This is handled in the `post-preview-button.js`:

```
// Request an autosave. This happens asynchronously and causes the component
		// to update when finished.
		if ( this.props.isDraft ) {
			this.props.savePost( { isPreview: true } );
		} else {
			this.props.autosave( { isPreview: true } );
		}
```
_is the autosave record "stale" when switching to draft?_
Yes, the autosave record that was previously generated for the published post remains in the Database, but additional edits to the draft post do not get reflected into the autosave record. Any new edits to the draft are reflected into the original record (the parent post). Reloading the page deletes the stale record.

Thanks,
- Jorge.

Reply:
Ok so is this correct?:

- the autosave exists because the post was first published and auto saves are only created for published posts.
- the post is switched to _Draft_ which causes autosaves **not to be updated** on further post content "change"
- when requesting a preview, because the auto-save _exists_ the code uses the preview link from that (stale) auto save
- the rendered preview is thus out of date with the current post content

So the solution is if the post is a draft don't use the autosave preview link from the autosave.

Reply:
Exactly. 

I have added a brief comment and a link to the PR and to the issue.